include/multipass/logging/log.h

```diff
@@ -43,24 +48,116 @@ namespace logging
  *
  * @tparam Arg0 Type of the first format argument
  * @tparam Args Type of the rest of the format arguments
- * @param level Log level
- * @param category Log category
- * @param fmt Format string
- * @param arg0 The first format argument
- * @param args Rest of the format arguments
+ * @param [in] level Log level
+ * @param [in] category Log category
+ * @param [in] fmt Format string
+ * @param [in] arg0 The first format argument
+ * @param [in] args Rest of the format arguments
  */
 template <typename Arg0, typename... Args>
 constexpr void
 log(Level level, const std::string& category, fmt::format_string<Arg0, Args...> fmt, Arg0&& arg0, Args&&... args)
 {
     const auto formatted_log_msg = fmt::format(fmt, std::forward<Arg0>(arg0), std::forward<Args>(args)...);
-    log(level, category, formatted_log_msg);
+    logging::log(level, category, formatted_log_msg);
+}
+
+/**
+ * Log function with templated log level.
+ *
+ * This function acts as an dispatch point for the templated
+ * and non-templated log overloads, based on argument count.
+ * We cannot simply use the templated overload since it requires
+ * at least 1 arguments.
+ *
+ * @tparam level Log level
+ * @tparam Args Format argument types
+ * @param [in] category Log category
+ * @param [in] fmt Format string
+ * @param [in] args Format arguments, if any
+ */
+template <Level level, typename... Args>
+constexpr void log(const std::string& category, fmt::format_string<Args...> fmt, Args&&... args)
+{
+    if constexpr (sizeof...(Args) > 0)
+    {
+        // Needs formatting
+        logging::log(level, category, fmt, std::forward<Args>(args)...);
+        return;
+    }
+    logging::log(level, category, CString{fmt.get().data()});
```

Comment:
I'm having second thoughts about this part. `.get()` returns a `fmt::string_view`, and `.data()` returns a `const char*` to a buffer, which is not guaranteed to be NUL-terminated. So, to ensure that the buffer is NUL-terminated, we have to construct a `std::string`. Alternatively, we can add a new log API function that can work with string views and handle them properly. I'm more inclined towards the second option.

Reply:
Maybe it's time to get rid of the `CString`.

Reply:
I vote to get rid of the `CString`

Reply:
> I vote to get rid of the `CString`

:genie:   https://github.com/canonical/multipass/pull/3969